### PR TITLE
Tdi operations and notifications

### PR DESCRIPTION
**Pull request by satish153**
notifications and operations support

### CODE REVIEW

File: include/tdi/common/tdi_notifications.hpp [hunk 1]
```diff
@@ -17,72 +17,83 @@
...
+                           std::unique_ptr<tdi::NotificationParams> params,
+                           void *cookie)>
+    tdiNotificationCallback;
+}  // namespace tdi
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+/**
+ * @brief TDI Notification c callback function
```
Comment (saynb) [inline include/tdi/common/tdi_notifications.hpp:88]: Should it be a separate C frontend file ?

File: src/c_frontend/tdi_table_notifications_c.cpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+                                      const uint64_t value) {
+  auto notification_param = reinterpret_cast<tdi::NotificationParams *>(notifications_hdl);
+  return notification_param->setValue(field_id, value);
+}
+
+// tdi_status_t tdi_notification_param_set_value_array(tdi_notification_param_hdl *data_hdl,
+//                                              const tdi_id_t field_id,
+//                                              const uint32_t *val,
+//                                              const uint32_t num_array) {
+//   auto notification_param = reinterpret_cast<tdi::NotificationParam *>(notification_hdl);
```
Comment (saynb) [inline src/c_frontend/tdi_table_notifications_c.cpp:39]: remove ?

File: src/c_frontend/tdi_table_notifications_c.cpp [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    uint32_t *array_size) {
+  auto notification_param = reinterpret_cast<const tdi::NotificationParams *>(notification_hdl);
+  std::vector<uint64_t> vec;
+  auto status = notification_param->getValue(field_id, &vec);
+  *array_size = vec.size();
+  return status;
+}
+
+#if 0 // TODO
+tdi_status_t tdi_notifications_set_value_ptr(tdi_notification_param_hdl *notifications_hdl,
```
Comment (saynb) [inline src/c_frontend/tdi_table_notifications_c.cpp:88]: maybe remove this code ?

File: src/c_frontend/tdi_table_notifications_c.cpp [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+tdi_status_t tdi_notifications_get_value_ptr(const tdi_notification_param_hdl *notifications_hdl,
+                                      tdi_notification_param_type_e type,
+                                           const size_t size,
+                                           uint8_t *val) {
+  auto notification_param = reinterpret_cast<const tdi::NotificationParams *>(notifications_hdl);
+  return notification_param->getValue(type, size, val);
+}
+#endif
```
Comment (saynb) [inline src/c_frontend/tdi_table_notifications_c.cpp:103]: Let's correct the EOF

File: include/tdi/common/tdi_notifications.hpp (hunk 1 above)
Comment (satish153) [reply include/tdi/common/tdi_notifications.hpp:88]: c callback signature is needed by notificationRegisterC in tdi_table.hpp. 
  virtual tdi_status_t notificationRegisterC(
      const tdi::Target &target,
      const tdi_id_t &notification_id,
      const tdi_notification_callback &callback_fn,
      const tdi::NotificationParams &registration_params,
      void *cookie) const;

File: src/c_frontend/tdi_table_notifications_c.cpp (hunk 2 above)
Comment (satish153) [reply src/c_frontend/tdi_table_notifications_c.cpp:39]: Done

File: src/c_frontend/tdi_table_notifications_c.cpp (hunk 3 above)
Comment (satish153) [reply src/c_frontend/tdi_table_notifications_c.cpp:88]: Done

File: src/c_frontend/tdi_table_notifications_c.cpp (hunk 4 above)
Comment (satish153) [reply src/c_frontend/tdi_table_notifications_c.cpp:103]: Done

File: src/c_frontend/tdi_table_operations_c.cpp [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+tdi_status_t tdi_operations_get_value_ptr(const tdi_operations_hdl *operations_hdl,
+                                      tdi_operations_field_type_e type,
+                                           const size_t size,
+                                           uint8_t *val) {
+  auto operations_field = reinterpret_cast<const tdi::TableOperations *>(operations_hdl);
+  return operations_field->getValue(type, size, val);
+}
+#endif
```
Comment (saynb) [inline src/c_frontend/tdi_table_operations_c.cpp:58]: Maybe correct EOF ?

File: src/tdi_table.cpp [hunk 6]
```diff
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
 #include <algorithm>
-#include <tdi/common/tdi_notifications.hpp>
+//#include <tdi/common/tdi_notifications.hpp>
```
Comment (saynb) [inline src/tdi_table.cpp:17]: remove?

File: tdi_python/tdiTable.py [hunk 7]
```diff
@@ -97,7 +97,9 @@ def __init__(self, cintf, handle, info, info_handle):
         self.supported_commands = ["info", "entry", "string_choices"]
         self.set_supported_attributes_to_supported_commands()
         self.set_supported_operations_to_supported_commands()
+        self.set_supported_notifications_to_supported_commands()
         self.set_supported_apis_to_supported_commands()
+        print("table: ", self.name, "supported: ", self.supported_commands)
```
Comment (saynb) [inline tdi_python/tdiTable.py:102]: Maybe use logging module and debug ?

File: tdi_python/tdiTable.py [hunk 8]
```diff
@@ -1802,11 +1804,40 @@ def set_supported_operations_to_supported_commands(self):
         keys_list = self.operations_type_cls.operations_dict.keys()
 
         # TODO: Enable this when operations are supported
-        '''
```
Comment (saynb) [inline tdi_python/tdiTable.py:1805]: Remove above TODO

File: tdi_python/tdiTable.py [hunk 9]
```diff
@@ -1802,11 +1804,40 @@ def set_supported_operations_to_supported_commands(self):
...
+            raise TdiTableError("Error: num notifications supported get failed on table {}. [{}]".format(self.name, self._cintf.err_str(sts)), self, sts)
+
+        arr_type = c_char_p * num_notifs.value
+        notifs_arr = arr_type()
+        sts = self._cintf.get_driver().tdi_table_notifications_supported(self._info_handle, byref(notifs_arr), byref(num_notifs))
+
+        for notif in notifs_arr:
+            self.supported_commands += [notif.decode('ascii')] #self.attributes_type_cls.attributes_dict[notifs_arr[i]]
+
+        print("supported_commands: ", self.supported_commands)
```
Comment (saynb) [inline tdi_python/tdiTable.py:1824]: logging module and remove commented code

File: include/tdi/common/c_frontend/tdi_operations.h [hunk 10]
```diff
@@ -24,7 +24,7 @@
 extern "C" {
 #endif
 
-tdi_status_t tdi_operations_set_value(tdi_operations_hdl operations,
+tdi_status_t tdi_operations_set_value(tdi_operations_hdl *operations,
```
[the remark this answers is not in the document]
Comment (satish153) [reply include/tdi/common/c_frontend/tdi_operations.h:27]: yeah, typo. Fixed

File: include/tdi/common/c_frontend/tdi_table.h [hunk 11]
```diff
@@ -22,9 +22,11 @@
 
 #include <tdi/common/c_frontend/tdi_attributes.h>
 #include <tdi/common/c_frontend/tdi_operations.h>
+#include <tdi/common/c_frontend/tdi_notifications.h>
 #include <tdi/common/c_frontend/tdi_table_data.h>
 #include <tdi/common/c_frontend/tdi_table_key.h>
 #include <tdi/common/tdi_defs.h>
+// #include <tdi/common/tdi_notifications.hpp>
```
[the remark this answers is not in the document]
Comment (satish153) [reply include/tdi/common/c_frontend/tdi_table.h:29]: Done

File: src/c_frontend/tdi_table_operations_c.cpp (hunk 5 above)
Comment (satish153) [reply src/c_frontend/tdi_table_operations_c.cpp:58]: Done

File: tdi_python/tdiTable.py (hunk 8 above)
Comment (satish153) [reply tdi_python/tdiTable.py:1805]: Done

File: tdi_python/tdiTable.py (hunk 9 above)
Comment (satish153) [reply tdi_python/tdiTable.py:1824]: Done

File: src/tdi_table.cpp (hunk 6 above)
Comment (satish153) [reply src/tdi_table.cpp]: done

File: tdi_python/tdiTable.py (hunk 7 above)
Comment (satish153) [reply tdi_python/tdiTable.py:102]: done

File: include/tdi/common/c_frontend/tdi_operations.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (saynb) [reply include/tdi/common/c_frontend/tdi_operations.h:27]: I can still see the getValue using non-ptr. Maybe it was missed

File: tdi_python/tdiTable.py (hunk 9 above)
Comment (saynb) [reply tdi_python/tdiTable.py]: I don't see this change reflected in the PR. Maybe it was missed?

**Review by saynb — APPROVED** [review]
LGTM. Please take care of the missed review comment and logging module usage in tdi_python instead of prints

File: include/tdi/common/c_frontend/tdi_operations.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (satish153) [reply include/tdi/common/c_frontend/tdi_operations.h:27]: Oh.. I missed pushing the latest commit.

**Pull request merged**